### PR TITLE
fix(database): Commence recovery if database is started with stale data container

### LINKED ISSUE #3234: deis limits:set not persistent after upgrade
I upgraded from `1.2.1` -> `1.4.1` and everything seems to be working fine except all of the custom limits I had set on a per-app basis had disappeared and now are 'unset' to `Unlimited`..

    [bcochran@schala a-demo|master *]$ deis releases
    === a-demo Releases
    v4      Sun Feb 15 10:49:51 2015 CST bcochran deployed b803d82
    v3      Mon Jan 26 16:10:29 2015 CST bcochran deployed 9cef9c1
    v2      Sun Jan 18 13:22:09 2015 CST bcochran deployed d328a95
    v1      Sun Jan 18 13:14:15 2015 CST bcochran created initial release
    
    [bcochran@schala a-demo|master *]$ deis limits
    === a-demo Limits
    
    --- Memory
    Unlimited
    
    --- CPU
    Unlimited

Just yesterday I set these to 712M and 1024 respectively.  It seems those revisions get wiped clean upon upgrade and do not get reset properly.

### LINKED ISSUE #3276: bug(database): wal-e rewrote backup
The node where the database was running was rebooted and after ir started in another node, all the apps (like 50) were not there.
This is the log:
```
Mar 11 20:55:21 nodo-2 sh[6169]: 2015-03-11T20:55:21Z 2e9ec5be7ae0 confd[49]: ERROR exit status 127
Mar 11 20:55:21 nodo-2 sh[6169]: database: waiting for confd to write initial templates...
Mar 11 20:55:21 nodo-2 sh[6169]: 2015-03-11T20:55:21Z 2e9ec5be7ae0 confd[49]: ERROR exit status 127
Mar 11 20:55:26 nodo-2 sh[6169]: database: existing data directory found. Starting postgres...
Mar 11 20:55:26 nodo-2 sh[6169]: 2015-03-11 20:55:26 UTC LOG:  could not open usermap file "/var/lib/postgresql/main/pg_ident.conf": No such file or directory
Mar 11 20:55:26 nodo-2 sh[6169]: 2015-03-11 20:55:26 UTC LOG:  database system was interrupted; last known up at 2015-03-11 20:54:35 UTC
Mar 11 20:55:26 nodo-2 sh[6169]: 2015-03-11 20:55:26 UTC LOG:  database system was not properly shut down; automatic recovery in progress
Mar 11 20:55:26 nodo-2 sh[6169]: 2015-03-11 20:55:26 UTC LOG:  redo starts at 71/D7000090
Mar 11 20:55:26 nodo-2 sh[6169]: 2015-03-11 20:55:26 UTC LOG:  record with zero length at 71/D701B858
Mar 11 20:55:26 nodo-2 sh[6169]: 2015-03-11 20:55:26 UTC LOG:  redo done at 71/D701B828
Mar 11 20:55:26 nodo-2 sh[6169]: 2015-03-11 20:55:26 UTC LOG:  last completed transaction was at log time 2015-03-11 20:54:45.108194+00
Mar 11 20:55:26 nodo-2 sh[6169]: 2015-03-11 20:55:26 UTC LOG:  autovacuum launcher started
Mar 11 20:55:26 nodo-2 sh[6169]: 2015-03-11 20:55:26 UTC LOG:  database system is ready to accept connections
Mar 11 20:55:27 nodo-2 sh[6169]: ALTER ROLE
Mar 11 20:55:27 nodo-2 sh[6169]: ALTER ROLE
Mar 11 20:55:27 nodo-2 sh[6169]: database: postgres is running...
Mar 11 20:55:27 nodo-2 sh[6169]: 2015-03-11 20:55:27 UTC ERROR:  role "deis" already exists
Mar 11 20:55:27 nodo-2 sh[6169]: 2015-03-11 20:55:27 UTC STATEMENT:  CREATE ROLE deis WITH LOGIN;
Mar 11 20:55:27 nodo-2 sh[6169]: ERROR:  role "deis" already exists
Mar 11 20:55:27 nodo-2 sh[6169]: 2015-03-11 20:55:27 UTC ERROR:  database "deis" already exists
Mar 11 20:55:27 nodo-2 sh[6169]: 2015-03-11 20:55:27 UTC STATEMENT:  CREATE DATABASE deis WITH OWNER deis;
Mar 11 20:55:27 nodo-2 sh[6169]: ERROR:  database "deis" already exists
Mar 11 20:56:27 nodo-2 sh[6169]: wal_e.main   INFO     MSG: starting WAL-E
Mar 11 20:56:27 nodo-2 sh[6169]: DETAIL: The subcommand is "wal-push".
Mar 11 20:56:27 nodo-2 sh[6169]: STRUCTURED: time=2015-03-11T20:56:26.820415-00 pid=287
Mar 11 20:56:27 nodo-2 sh[6169]: wal_e.worker.upload INFO     MSG: begin archiving a file
Mar 11 20:56:27 nodo-2 sh[6169]: DETAIL: Uploading "pg_xlog/0000001600000071000000D7" to "s3://db_wal/wal_005/0000001600000071000000D7.lzo".
Mar 11 20:56:27 nodo-2 sh[6169]: STRUCTURED: time=2015-03-11T20:56:26.846134-00 pid=287 action=push-wal key=s3://db_wal/wal_005/0000001600000071000000D7.lzo prefix= seg=0000001600000071000000D7 state=begin
Mar 11 20:56:27 nodo-2 sh[6169]: wal_e.worker.upload INFO     MSG: completed archiving to a file
Mar 11 20:56:27 nodo-2 sh[6169]: DETAIL: Archiving to "s3://db_wal/wal_005/0000001600000071000000D7.lzo" complete at 7326.81KiB/s.
Mar 11 20:56:27 nodo-2 sh[6169]: STRUCTURED: time=2015-03-11T20:56:27.086045-00 pid=287 action=push-wal key=s3://db_wal/wal_005/0000001600000071000000D7.lzo prefix= rate=7326.81 seg=0000001600000071000000D7 state=complete
```

### PR DESCRIPTION
**Pull request by romansergey**
I popped up with this on IRC on Friday, but to recap, the fail scenario is as follows:

1. start db on node A
2. force reschedule to node B
3. perform some updates
4. force reschedule to node A

Result: db has started successfully on node A with an existing, outdated `deis-database-data` container and is serving old data.

The PR provides a fix for that with a test script.

Note on PR:

* The test takes twice as long to complete compared to `database_test.go`, but it is just as isolated, so I've placed it in `database/tests`.
* I had to patch Mock-S3 in mock-store for wal-e not to frustrate over inability to delete old wal segments. I'm not sure what is the best way to keep it until upstream is updated.

// edit by @carmstrong to add referenced tickets
fixes #3276
fixes #3234

### CODE REVIEW

**Comment by deis-admin** [comment]
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

**Comment by mboersma** [comment]
Jenkins, add to whitelist.

**Comment by carmstrong** [comment]
This is lovely. Thank you so much, @romansergey! I'll want to do some hand testing of this.

**Comment by romansergey** [comment]
My pleasure, @carmstrong.

The test script is not the DRY-est I ever wrote, but to help, the high-level script:

1. Create blank volumes `A` and `B`
2. Run database container with volume `A` and wait for it to start accepting clients.
3. Shutdown
4. Run database container with volume `B` and wait for it to start accepting clients.
5. Create table (for the sake of having a verifiable update)
6. Wait for the next backup cycle to start and complete
7. Shutdown
8. Run database container again with volume `A` and wait for it to start accepting clients.
9. Assert the table existence

**Comment by aledbf** [comment]
@carmstrong Why not just remove the data container and recover the log from scratch?

**Comment by carmstrong** [comment]
> Why not just remove the data container and recover the log from scratch?

Because that takes too long for a simple restart of the component, which is a common case. We want speedy recovery here, whatever we have to do to achieve that.

**Comment by carmstrong** [comment]
We need to test this on upgrading clusters, too, since they won't have an initialize ID but their data is current.

**Comment by bacongobbler** [comment]
> Because that takes too long for a simple restart of the component, which is a common case. We want speedy recovery here, whatever we have to do to achieve that.

Given the following scenario:

 - start db on node A
 - create application
- force reschedule to node B
- delete app
- force reschedule to node A

What happens? Will wal-e know how to recover? I'm personally in favour of removing the data container and recovering from scratch if it means recoveries are consistent.

**Comment by carmstrong** [comment]
> What happens? Will wal-e know how to recover?

The issue here isn't with wal-e at all - the issue is with the database's /bin/boot logic, which will only perform a restore/replay if it detects that it has no data. If the database comes up on a host which previously had the database running, it happily starts serving "old" data, even though newer backup data exists.

**Comment by aledbf** [comment]
> which will only perform a restore/replay if it detects that it has no data.. If the database comes up on a host which previously had the database running, it happily starts serving "old" data...

That makes sense. Is what I saw in #3276. 
After the boot process one of the things is to upload a new backup and that upload contains "old" data (but wal-e is not aware of that)

**Comment by carmstrong** [comment]
> After the boot process one of the things is to upload a new backup and that upload contains "old" data (but wal-e is not aware of that)

Yep. If we can ensure that wal-e always has the latest, I think that problem goes away.

**Comment by ianblenke** [comment]
Now I think I understand the problem. Yes, I'm all for destroying old deis-database-data containers when deis-database moves to another fleet machine. If we're relying on ceph for persistence, we can't trust it half-way. All-in.

**Comment by carmstrong** [comment]
@romansergey Can you rebase this off master? I want to ensure it passes with recent test changes (and we bumped wal-e to fix a build issue).

**Comment by deis-admin** [comment]
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

**Comment by romansergey** [comment]
@carmstrong, done.

I agree with the other guys on the simple container deletion for as long as bare stop/start of the database service will not wipe it and trigger full recovery from the wal.
If very few (as far as I can see) experienced this issue by this point, that means very few will benefit from the commenced recovery.

I think these changes will definitely have more effect in combination with something like scheduling an extra warm-standby database service.

**Comment by carmstrong** [comment]
> I agree with the other guys on the simple container deletion for as long as bare stop/start of the database service will not wipe it and trigger full recovery from the wal.

It will, unfortunately. There's no way to tell fleet to purge the data container only when a unit is rescheduled -- it either happens never or on every start. I think that removing the data containers is a no-go, as it means both restarting the unit and a normal failover would take several minutes. 

This PR is a lot of excellent work and fixes an obvious error case, so it's something we should merge.

**Comment by gabrtv** [comment]
> I think that removing the data containers is a no-go, as it means both restarting the unit and a normal failover would take several minutes.

Agreed.  The data container is an important optimization for the reasons you mentioned.  

We need to put some more thought into how we can facilitate a warm-standby database that provides quick failover without requiring WAL replay on boot.  It's definitely achievable with leader election.

> This PR is a lot of excellent work and fixes an obvious error case, so it's something we should merge.

:+1:

**Comment by romansergey** [comment]
The latest test job (https://ci.deis.io/job/test-ec2/146/) failed due to missing changes in respect to the new image name prefixing. Those have been added now.

Please re-run the tests.

**Comment by carmstrong** [comment]
> Please re-run the tests.

Running now: https://ci.deis.io/job/test-ec2/153/

**Comment by romansergey** [comment]
@carmstrong, that ain't me! :scream_cat:

 https://ci.deis.io/job/test-ec2/153/consoleFull#682192135502066c60b2236c09cc9bf8

**Comment by romansergey** [comment]
The only assumption I have is that it stalled in a connection attempt with no set timeout in [the latest job](https://ci.deis.io/job/test-ec2/158/consoleFull#682192135502066c60b2236c09cc9bf8). That would be fair.

@carmstrong, I have amended the relevant changes - please re-run the tests.

**Comment by bacongobbler** [comment]
https://ci.deis.io/job/test-ec2/212/ will start once a slave is freed up

File: Godeps/Godeps.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
 	"ImportPath": "github.com/deis/deis",
-	"GoVersion": "go1.4.1",
+	"GoVersion": "go1.2.1",
```
Comment (bacongobbler) [inline Godeps/Godeps.json]: would you mind updating this back to v1.4.1?

**Comment by bacongobbler** [comment]
Code LGTM otherwise

**Comment by mboersma** [comment]
@romansergey, the cleanup commits like `ref(tests) obey the gofmt-lord` and `ref(tests): please gofmt over new changes` and probably a couple others should be [squashed into neighboring commits](http://docs.deis.io/en/latest/contributing/standards/#single-issue):

> Squash and rebase the commit or commits in your pull request into logical units of work with git.

Thanks again!

**Comment by romansergey** [comment]
@mboersma , you've got it

**Comment by romansergey** [comment]
@mboersma , @bacongobbler now it is done. Please test.

**Comment by mboersma** [comment]
Awesome, thank you. I will test it now.

**Comment by mboersma** [comment]
This is looking good so far. I'm going to test an upgrade from an earlier version of Deis and see how that goes.

**Comment by mboersma** [comment]
Tested on vagrant, LGTM.

**Comment by carmstrong** [comment]
@romansergey The mock-s3 patch no longer applies, so we reverted that in #4800. Can you recall why that was necessary, and how we can validate that removing it didn't break anything?

Thanks!

**Comment by krancour** [comment]
@carmstrong nothing to validate anymore.  I've pinned us to an older version of mock-s3 and kept the patch.

**Pull request merged**